Dwell graduation - Adding volumetric UI examples & polish

cre8ivepark opened the pull request:
## Overview
The current Dwell Example scene only shows Unity UI examples.
Added volumetric UI examples & polish. Added prefab variants of Pressable HoloLens 2 Buttons that have additional quad and scripts that visualize dwell by modifying _BorderWidth property of MRTK Standard Shader.

<img width="965" alt="2021-05-10 11_21_18-Window" src="https://user-images.githubusercontent.com/13754172/117713159-250acc80-b18a-11eb-835c-8a198a76d910.png">

https://user-images.githubusercontent.com/13754172/117713133-1cb29180-b18a-11eb-9a32-fba86c4bce4e.mp4


## Added files
**BaseDwellPressableButton**.cs : Variation of BaseDwellSample.cs that works with PressableButtonHoloLens2 components
**ToggleDwellPressableButton**.cs : Variation of ToggleDwellSample.cs that works with PressableButtonHoloLens2 components
Prefab variants of volumetric HoloLens2 buttons. These prefabs have an additional quad to visualize the dwell.
**PressableButtonHoloLens2ToggleDwell**.prefab
**PressableButtonHoloLens2Toggle_32x96_Dwell**.prefab
**PressableButtonHoloLens2_32x96_Dwell**.prefab
**PressableButtonHoloLens2_Dwell**.prefab 

These two scripts were added to the prefab variants:

<img width="251" alt="2021-05-10 12_25_20-Window" src="https://user-images.githubusercontent.com/13754172/117713720-e0cbfc00-b18a-11eb-9850-6f1eeedcc01b.png">
<img width="279" alt="2021-05-10 12_25_29-MRTK - DwellExample - Universal Windows Platform - Unity 2019 4 22f1  PREVIEW PA" src="https://user-images.githubusercontent.com/13754172/117713727-e295bf80-b18a-11eb-9fa2-1ce0bfc69afa.png">

RogPodge commented:
Tested out the scene, some of the button prefabs don't seem to have the Button Background field assigned, resulting in null reference exceptions.

![image](https://user-images.githubusercontent.com/39840334/117895610-91152f80-b273-11eb-84d3-c29d99f04514.png)

Is there a way we can alternate default behaviors for when these fields aren't set. It would help users be able to take advantage of these new components much more easily

RogPodge commented:
It's also not clear that the Toggle Button being 'On' is what allows the 'Load More' buttons to actually load more options.

Additionally, do we need 2 load more options? I feel like one should be "load previous options" and the other should be "load next options" from a intuitive functionality standpoint

![image](https://user-images.githubusercontent.com/39840334/117895886-35977180-b274-11eb-9d68-2a9ea4342ba7.png)

RogPodge commented on Assets/MRTK/Examples/Demos/UX/Dwell/ToggleDwellPressableButton.cs:
```diff
@@ -0,0 +1,103 @@
+        {
+            base.DwellCompleted(pointer);
+            buttonBackground.GetComponent<MeshRenderer>().material.color = isDwellEnabled ? this.dwellOnColor : this.dwellOffColor;
+            //dwellVisualImage.transform.localScale = Vector3.zero;
```
```suggestion
```

RogPodge commented on Assets/MRTK/Examples/Demos/UX/Dwell/ToggleDwellPressableButton.cs:
```diff
@@ -0,0 +1,103 @@
+        private TextMeshPro dwellStatus = null;
+
+        [SerializeField]
+        private GameObject buttonBackground = null;
```
Can we make this of type MeshRenderer, since all instance of this object invoke GetComponent<MeshRenderer>()?

RogPodge commented on Assets/MRTK/Examples/Demos/UX/Dwell/BaseDwellPressableButton.cs:
```diff
@@ -0,0 +1,72 @@
+    public abstract class BaseDwellPressableButton : MonoBehaviour
+    {
+        [SerializeField]
+        protected GameObject dwellVisualImage = null;
```
Can we make this of type MeshRenderer, since all instance of this object invoke GetComponent<MeshRenderer>()?

cre8ivepark replied:
That makes sense!

cre8ivepark replied:
Oh, I found that there are some lines that use transform.

RogPodge replied:
You can still use the .transform of a MeshRenderer and it will change the transform of the gameobject that owns the component.

cre8ivepark commented:
> It's also not clear that the Toggle Button being 'On' is what allows the 'Load More' buttons to actually load more options.
> 
> Additionally, do we need 2 load more options? I feel like one should be "load previous options" and the other should be "load next options" from a intuitive functionality standpoint
> 
> ![image](https://user-images.githubusercontent.com/39840334/117895886-35977180-b274-11eb-9d68-2a9ea4342ba7.png)

@RogPodge Actually, I haven't modified this part (Unity UI examples) created by Ritika but totally agree. Some buttons and results are vague. It seems those two buttons show slightly different dwell decay behavior. I'll look into it. Maybe we can add more clear descriptions.

kndrewa commented:
This is fantastic work.  One of the most important concepts in d365MR apps (for example, hands free workflows in d365Guides) is dynamic audio parameters and the associated audio manager infrastructure.  d365MR would like to request this in MRTK dwell to be able to abandon our custom solution (since sonic feedback is critical to the user understanding their success/fail state and expression of immediacy/success on activation).  We use a duration parameter to change pitch and tone in the sound as the action progresses.

cre8ivepark commented:
@RogPodge Made some updates. Removed the example code in the ToggleDwellSample.cs (and ToggleDwellPressableButton.cs) which searches and disables all DwellHandlers. I don't think this is a good example and makes it confusing. Visualizing the toggle state would be enough for the toggle button demo. Now list view and two load buttons are working regardless of the toggle button state. Added two text label to describe the difference between two load more buttons.

![2021-05-14 20_14_11-MRTK-Public - DwellExample - Universal Windows Platform - Unity 2019 3 13f1 Pers](https://user-images.githubusercontent.com/13754172/118346532-23763700-b4f1-11eb-9478-883804ce9baf.png)

https://user-images.githubusercontent.com/13754172/118346553-53253f00-b4f1-11eb-9657-0ee2215ba58a.mp4

cre8ivepark commented:
/azp run

RogPodge commented:
the CI seems to have had some issues with the dependency changes made to the prefab:

ERROR: PressableButtonHoloLens2Toggle_32x96_Dwell.prefab is NOT allowed to depend upon files in the Examples package.
 @ D:\a\1\s\Assets\MRTK\Examples\Demos\UX\Dwell\DwellProfileWithDecay.asset
Examples : D:/a/1/s/Assets/MRTK/Examples/Demos/UX/Dwell/DwellProfileWithDecay.asset

ERROR: PressableButtonHoloLens2Toggle_32x96_Dwell.prefab is NOT allowed to depend upon files in the Examples package.
 @ D:\a\1\s\Assets\MRTK\Examples\Demos\UX\Dwell\ToggleDwellPressableButton.cs
Examples : D:/a/1/s/Assets/MRTK/Examples/Demos/UX/Dwell/ToggleDwellPressableButton.cs

ERROR: PressableButtonHoloLens2Toggle_32x96_Dwell.prefab is NOT allowed to depend upon files in the Examples package.
 @ D:\a\1\s\Assets\MRTK\SDK\Features\UX\Interactable\Prefabs\PressableButtonHoloLens2Toggle_32x96.prefab
Foundation : D:/a/1/s/Assets/MRTK/SDK/Features/UX/Interactable/Prefabs/PressableButtonHoloLens2Toggle_32x96.prefab
 @ D:\a\1\s\Assets\MRTK\StandardAssets\Audio\MRTK_Select_Main.wav
StandardAssets : D:/a/1/s/Assets/MRTK/StandardAssets/Audio/MRTK_Select_Main.wav
 @ D:\a\1\s\Assets\MRTK\SDK\Features\UX\Interactable\Scripts\Interactable.cs
Foundation : D:/a/1/s/Assets/MRTK/SDK/Features/UX/Interactable/Scripts/Interactable.cs
ERROR: D:\a\1\s\Assets\MRTK\SDK\Features\UX\Interactable\Prefabs\PressableButtonHoloLens2Toggle_32x96_Dwell.prefab contains 3 dependency errors.

RogPodge commented:
/azp run

RogPodge commented:
Fixed up the rest of the dwell scene. Since the scene seems to have been initially made in Unity 2019, many of the button visuals won't look quite right in Unity 2018

Also added unit tests for the dwell cancel state and the dwell decay settings

RogPodge commented:
/azp run

cre8ivepark commented:
Looks good on the device. Slightly adjusted content z position to make it easy to interact with gaze.

cre8ivepark commented:
/azp run
